vault/data_source_generic_secret_list_test.go

```diff
@@ -0,0 +1,141 @@
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+
+	"github.com/hashicorp/terraform-provider-vault/testutil"
+)
+
+func TestDataSourceGenericSecretList_v1(t *testing.T) {
+	mount := acctest.RandomWithPrefix("tf-kv")
+	s1 := acctest.RandomWithPrefix("foo")
+	s2 := acctest.RandomWithPrefix("bar")
+
+	resourceName := "data.vault_generic_secret_list.test"
+	resource.Test(t, resource.TestCase{
+		Providers: testProviders,
+		PreCheck:  func() { testutil.TestAccPreCheck(t) },
+		Steps: []resource.TestStep{
+			{
+				Config: testDataSourceGenericSecretListConfig(mount, s1, s2, false),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "path", mount),
+					resource.TestCheckResourceAttr(resourceName, "names.#", "3"),
+					resource.TestCheckResourceAttr(resourceName, "names.0", s2),
+					resource.TestCheckResourceAttr(resourceName, "names.1", fmt.Sprintf("%s/", s2)),
+					resource.TestCheckResourceAttr(resourceName, "names.2", s1),
+				),
+			},
+		},
+	})
+}
+
+func TestDataSourceGenericSecretList_v2(t *testing.T) {
+	mount := acctest.RandomWithPrefix("tf-kv")
+	s1 := acctest.RandomWithPrefix("foo")
+	s2 := acctest.RandomWithPrefix("bar")
+
+	resourceName := "data.vault_generic_secret_list_v2.test"
+	resource.Test(t, resource.TestCase{
+		Providers: testProviders,
+		PreCheck:  func() { testutil.TestAccPreCheck(t) },
+		Steps: []resource.TestStep{
+			{
+				Config: testDataSourceGenericSecretListConfig(mount, s1, s2, true),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr(resourceName, "path", fmt.Sprintf("%s/metadata", mount)),
+					resource.TestCheckResourceAttr(resourceName, "names.#", "3"),
+					resource.TestCheckResourceAttr(resourceName, "names.0", s2),
+					resource.TestCheckResourceAttr(resourceName, "names.1", fmt.Sprintf("%s/", s2)),
+					resource.TestCheckResourceAttr(resourceName, "names.2", s1),
+				),
+			},
+		},
+	})
+}
+
+func testDataSourceGenericSecretListMountConfig(mount string, isV2 bool) string {
+	ret := fmt.Sprintf(`
+resource "vault_mount" "test" {
+	  path = "%s"
+	  type = "kv"
+	  description = "Example KV Mount."
+`, mount)
+	if isV2 {
+		ret += fmt.Sprintf(`
+	  options = {
+		  version = "2"
+	  }
+}`)
+	} else {
+		ret += fmt.Sprintf(`
+	  options = {
+		  version = "1"
+	  }
+}`)
+	}
+
+	return ret
+}
+
+func testDataSourceGenericSecretConfig(mount, secretPath1, secretPath2 string, isV2 bool) string {
+	return fmt.Sprintf(`
+%s
+
+resource "vault_generic_secret" "test_1" {
+    path      = "${vault_mount.test.path}/%s"
+    data_json = <<EOT
+{
+    "zip": "zap"
+}
+EOT
+}
+
+resource "vault_generic_secret" "test_2" {
+  path      = "${vault_mount.test.path}/%s"
+
+  data_json = <<EOT
+{
+  "foo":   "bar",
+  "pizza": "cheese"
+}
+EOT
+}
+
+resource "vault_generic_secret" "test_nested" {
+  path = "${vault_generic_secret.test_2.path}/biz"
+
+  data_json = <<EOT
```

Comment:
I  prefer to to `jsonencode()` instead of the HERE doc syntax.

Reply:
Updated in latest commits